sources/api/apiclient/src/exec.rs

```diff
@@ -0,0 +1,683 @@
+                        debug!("Finished reading input");
+                        // If we can, send a ContentComplete message to the server so we can exit more
+                        // cleanly, but either way we're done.  (Again, we shouldn't get here.)
+                        if let Ok(msg) = serde_json::to_string(&ClientMessage::ContentComplete) {
+                            let _ = tx.unbounded_send(Message::Text(msg));
+                        } else {
+                            warn!("Unable to send content complete to server, may hang if process doesn't exit");
+                        }
+                        return Ok(());
+                    } else {
+                        // Any error other than EOF is a real read error.
+                        Err(e).context(error::ReadFromUser)?
+                    }
+                }
+            }
+        }
+    }
+
+    /// Read from stdin in bulk, sending larger batches of data at a time.
+    fn read_stdin(tx: mpsc::UnboundedSender<Message>, capacity: Arc<AtomicCapacity>) -> Result<()> {
+        let mut stdin = std::io::stdin();
+        // Keep track of the number of messages we've read.  We compare this to the number of
+        // messages the server has written, as received in its regular capacity update messages, so
+        // that we don't overwhelm the server.
+        let mut messages_read = 0u64;
+
+        loop {
+            // Wait for server to have capacity for writes before reading; don't give "false hope" to
+            // whatever's writing to our stdin that it'll be read until there's room for it.
+            Self::wait_for_capacity(messages_read, &capacity);
+
+            // Read a batch of data at a time; 4k is a balanced number for small and large jobs.
+            let mut buf = [0; 4096];
+            let count = stdin.retry_read(&mut buf).context(error::ReadFromUser)?;
+            // A read of 0 indicates EOF, so we're done.
+            if count == 0 {
+                break;
+            }
+            messages_read += 1;
+
+            // Send the data to the server in a Binary message without encoding.
+            let msg = Vec::from(&buf[..count]);
+            tx.unbounded_send(Message::Binary(msg))
+                .context(error::SendMessage { kind: "user input" })?;
+        }
+        debug!("Finished reading input");
+
+        // Send a ContentComplete message to the server so it can exit the process more cleanly.
+        // This is more important than the TTY case; interactive use typically has users typing
+        // exit, or quit, or ctrl-d... noninteractive programs typically wait for EOF.
+        let msg =
+            serde_json::to_string(&ClientMessage::ContentComplete).context(error::Serialize)?;
+        tx.unbounded_send(Message::Text(msg))
+            .context(error::SendMessage {
+                kind: "content complete",
+            })?;
+
+        Ok(())
+    }
```

Comment:
I'm not sure I follow.  In this function, serialization and send errors are handled in the same way - stopping with an error.  Neither should ever happen, and I've never seen them happen.  If send error fails here, though, it's more important than the TTY case because ContentComplete is more important than in the TTY case.  Noninteractive processes will generally hang if stdin isn't closed.  In the TTY case, generally stdin is closed by the user exiting a program, meaning the program doesn't need a "stdin is closed" signal to exit, it was told to exit.

JSON ContentComplete could perhaps be prepared in a lazy_static?  I guess it would have to panic if it couldn't be constructed, and then the error wouldn't be very clear.  A helper seems like a bit of overkill for this, to me, because it's short, and because it obscures the pattern being used for control messages, which I think is important to understand.  What would you have it look like?

Reply:
This makes more sense to me in the new revision, now that the serialization and send errors are handled consistently within `read_stdin_tty` and `read_stdin`, but differently between the two - the former warns in both cases, and the latter returns an error in both cases.